File: src/temp/ast/expression/Logical.java
package temp.ast.expression;

import temp.ast.Expression;
import temp.lang.Token;
import temp.lang.Visitor;

public class Logical extends Expression {
    private final Expression left;
    private final Token operator;
    private final Expression right;

    public Logical(Expression left, Token operator, Expression right) {
        this.left = left;
        this.operator = operator;
        this.right = right;
    }

    public <R> R accept(Visitor<R> visitor) {
        return visitor.visitLogicalExpression(this);
    }
}

File: src/codepulsar/ast/expression/Logical.cpp
#include "Logical.h"


Pulsar::Logical::Logical(Expression* left, std::string operatorType, Expression* right, int line) {
    this->left = left;
    this->operatorType = operatorType;
    this->right = right;
    this->line = line;
}

template<typename R>
R Pulsar::Logical::accept(ExprVisitor<R>& visitor) {
    visitor.visitLogicalExpression(this);
}

Pulsar::Expression* Pulsar::Logical::getLeft() {
    return this->left;
}

std::string Pulsar::Logical::getOperator() {
    return this->operatorType;
}

Pulsar::Expression* Pulsar::Logical::getRight() {
    return this->right;
}

int Pulsar::Logical::getLine() {
    return this->line;
}
